fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneSpellcheckRecordCursor.java

```diff
@@ -0,0 +1,234 @@
+                RecordCursorProto.LuceneSpellcheckIndexContinuation.newBuilder().setValue(ByteString.copyFromUtf8(lookupResult.toString()));
+        continuationBuilder.setLocation(currentPosition);
+        return ByteArrayContinuation.fromNullable(continuationBuilder.build().toByteArray());
+    }
+
+    @Override
+    public void close() {
+    }
+
+    @Nonnull
+    @Override
+    public Executor getExecutor() {
+        return executor;
+    }
+
+    @Override
+    public boolean accept(@Nonnull final RecordCursorVisitor visitor) {
+        visitor.visitEnter(this);
+        return visitor.visitLeave(this);
+    }
+
+    private synchronized IndexReader getIndexReader() throws IOException {
+        IndexWriterCommitCheckAsync writerCheck = getIndexWriterCommitCheckAsync(state, groupingKey);
+        return writerCheck == null ? DirectoryReader.open(getOrCreateDirectoryCommitCheckAsync(state, groupingKey).getDirectory()) : DirectoryReader.open(writerCheck.indexWriter);
+    }
+
+    private void spellcheck() throws IOException {
+        if (spellcheckSuggestions != null) {
+            return;
+        }
+        long startTime = System.nanoTime();
+        IndexReader indexReader = getIndexReader();
+
+        List<Suggestion> suggestionResults = new ArrayList<>();
+        for (String field : fields) {
+            //collect all suggestions across all given fields or across all fields depending on user options.
+            Arrays.stream(spellchecker.suggestSimilar(new Term(field, wordToSpellCheck), limit, indexReader))
+                    .map(suggestion -> new Suggestion(field, suggestion))
+                    .forEach(suggestionResults::add);
+        }
+        spellcheckSuggestions = suggestionResults.stream()
+                //Merge matching suggestions from different
+                .collect(Collectors.toMap(
+                        suggestion -> suggestion.suggestWord.string,
+                        Function.identity(),
+                        // TODO: For arnaud, are we checking for a merge on ALL suggested words against eachother?
+                        LuceneSpellcheckRecordCursor::mergeTwoSuggestWords))
+                .values()
+                .stream()
+                // Sort the suggested words from large to small by score then by frequency then by the field.
+                .sorted(comparing((Suggestion s) -> s.suggestWord.score).reversed()
+                                .thenComparing(comparing((Suggestion s) -> s.suggestWord.freq).reversed())
+                                .thenComparing(s -> s.suggestWord.string))
+                .limit(limit)
+                // Map the words from suggestions to index entries.
+                .map(suggestion -> new IndexEntry(
+                        state.index,
+                        Tuple.from(groupingKey.isEmpty() ? null : groupingKey.get(0).toString(), suggestion.indexField, suggestion.suggestWord.string),
```

Comment:
`groupingKey` can be anything; I don't see that it has to just be one-long or a string. Can just use `addAll`, I think.

Reply:
This seems to still be the same in the code that got merged, although it moved a few lines away.

Reply:
yes. If this is empty then addAll becomes a two item tuple which messes up our expected count for the returned indexEntry. I tried a few different things but this made the most sense.

Reply:
Wouldn't something like
```
groupingKey.addAll(Tuple.from(suggestion.indexField, suggestion.suggestWord.string))
```
be more correct? Or do I misunderstand the requirements?

Reply:
If the tuple is empty then it only has those two elements instead of the 3 it has if there’s a grouping key. The grouping key is first and if we have a variance of 2 or 3 elements in the tuple I assumed it would be harder to incorporate. It also maintains consistency with Yizuo’s auto complete.

Reply:
Hmm, okay. This gets cleaned up with the scan bounds changes, which should eliminate all this tuple packing and unpacking.